packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,123 @@
+    const [pair, associationKey] = this.associationPrecheck(scope, product);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey).get(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptLanguage,
+        description: options.description,
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        tagUpdateOnProvisionedProduct: options.tagUpdateOnProvisionedProductAllowed || options.tagUpdateOnProvisionedProductAllowed === undefined
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const tO = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, tO);
+        }
+
+        new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+          resourceId: resourceId,
+          tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
+        });
+      });
+    });
+  }
+
+  private static associationMap = new Map<string, any>();
+  private static tagOptionMap = new Map<string, CfnTagOption>();
+
+  private static associationPrecheck(scope: cdk.Resource, product: IProduct): [AssociationPair, string] {
+    const pair = this.resolveProductAndPortfolio(scope, product);
+    const associationKey = hashValues(pair.portfolio.node.addr, pair.product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      this.associateProductWithPortfolio(scope, pair.portfolio, pair.product);
+    }
+    return [pair, associationKey];
+  }
+
+  private static resolveProductAndPortfolio(scope: cdk.Resource, product: IProduct): AssociationPair {
+    return {
+      portfolio: scope as Portfolio,
+      product: product as Product,
```

Comment:
Sorry, what is happening here?

Reply:
A part of our roadmap is to make adding associations/constraints bidirectional, so you could do something like this: 
```
portfolio.addConstraint(product);
product.addConstraint(portfolio);
```
And those would call the same underlying API. In order to simplify this we either need to be able to resolve who is calling the method.  From what we've tried, we can either do it like here, or always explicitly pass in both portfolio and product, but that looks like ```(this, this, product, options?)``` which seemed more confusing.  If there is a better way to do this would be very interested.

Reply:
I meant more about the type casting. If those are the types you need, then you should require those types, not do casts that will result in runtime failures (think about what would happen if you used  `Portfolio` and `Product` imported with the `from*()` methods here).

Reply:
So this actually is a simplified version of what we had previously that was built for features like bidirectional capabilities.  Originally this method resolved by checking for `portfolioArn` in the scope, and then `productArn` in the future if necessary and using that to assign it, since we could use the scope to figure out what is calling it.  Do you think it is better to just use what I mentioned above where we always pass in portfolio, product on top of scope?  That way we never have to resolve anything, it just looks a little strange.

Reply:
Also these methods should work for imported resources.  Is that issue though since I am guessing for example that for imported issues there are deploy time failures that we cannot catch, for example if  a constraint already exists?

Reply:
To be more definitive, is this a better approach?
```
// in portfolio.ts
  public addResourceUpdateConstraint(product: IProduct, options?) {
    AssociationManager.addResourceUpdateConstraint(this, this, product, options);
  }
```
and then 
```
// association manager
public static addResourceUpdateConstraint(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct, options: TagUpdatesOptions)
```

Reply:
Got rid of this.  Punting this to future if we ever add bidirectional capabilities it can be addressed then.

Reply:
I have to say I don't quite understand the problem. Either you need `Product` and `Portfolio` there, in which case those are the types you should put in the signature of the methods, or it also works with `IPortfolio` and `IProduct`, in which case these casts are not needed, and can be removed.

Reply:
these have been removed in favor of just putting types in the method signature